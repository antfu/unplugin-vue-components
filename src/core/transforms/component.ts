import Debug from 'debug'
import type { CallExpression } from '@babel/types'
import type MagicString from 'magic-string'
import { pascalCase, stringifyComponentImport } from '../utils'
import type { Context } from '../context'

const debug = Debug('unplugin-vue-components:transform:component')

export default async(nodes: CallExpression[], version: 'vue2' | 'vue3', s: MagicString, ctx: Context, sfcPath: string) => {
  let no = 0

  for (const node of nodes) {
    const { callee, arguments: args } = node
    if (callee.type !== 'Identifier' || callee.name !== (version === 'vue2' ? '_c' : '_resolveComponent') || args[0].type !== 'StringLiteral')
      continue
    const componentName = args[0].value
    if (!componentName || componentName.startsWith('_'))
      continue

    debug(`| ${componentName}`)
    const name = pascalCase(componentName)
    ctx.updateUsageMap(sfcPath, [name])

    const component = await ctx.findComponent(name, 'component', [sfcPath])
    if (component) {
      const var_name = `__unplugin_components_${no}`
      s.prepend(`${stringifyComponentImport({ ...component, name: var_name }, ctx)};\n`)
      no += 1

      const replacedNode = version === 'vue2' ? args[0] : node
      s.overwrite(replacedNode.start!, replacedNode.end!, var_name)
    }
  }

  debug(`^ (${no})`)
}
