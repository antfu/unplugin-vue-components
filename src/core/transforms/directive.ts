import Debug from 'debug'
import type {
  CallExpression, ObjectProperty, File, VariableDeclaration, FunctionExpression, BlockStatement,
} from '@babel/types'
import type MagicString from 'magic-string'
import { parse, ParseResult } from '@babel/parser'
import traverse from '@babel/traverse'
import { pascalCase, stringifyComponentImport } from '../utils'
import type { Context } from '../context'
import { ResolveResult } from '../transformer'
import { VueVersion } from '../..'

const debug = Debug('unplugin-vue-components:transform:directive')

/**
 * get Vue 2 render function position
 * @param ast
 * @returns
 */
const getRenderFnStart = (ast: ParseResult<File>): number => {
  const renderFn = ast.program.body.find((node): node is VariableDeclaration =>
    node.type === 'VariableDeclaration'
    && node.declarations[0].id.type === 'Identifier'
    && node.declarations[0].id.name === 'render',
  )
  const start = (((renderFn?.declarations[0].init as FunctionExpression).body) as BlockStatement).start
  if (start === null)
    throw new Error('[unplugin-vue-components:directive] Cannot find render function position.')
  return start + 1
}

const resolveVue2 = (code: string, s: MagicString): ResolveResult[] => {
  const ast = parse(code, {
    sourceType: 'module',
  })
  const nodes: CallExpression[] = []
  traverse(ast, {
    CallExpression(path) {
      nodes.push(path.node)
    },
  })

  const results: ResolveResult[] = []
  for (const node of nodes) {
    const { callee, arguments: args } = node
    // _c(_, {})
    if (callee.type !== 'Identifier' || callee.name !== '_c' || args[1].type !== 'ObjectExpression')
      continue

    // { directives: [] }
    const directives = args[1].properties.find(
      (property): property is ObjectProperty =>
        property.type === 'ObjectProperty'
        && property.key.type === 'Identifier'
        && property.key.name === 'directives',
    )?.value
    if (!directives || directives.type !== 'ArrayExpression')
      continue

    const renderStart = getRenderFnStart(ast)

    for (const directive of directives.elements) {
      if (directive?.type !== 'ObjectExpression') continue

      const nameNode = directive.properties.find(
        (p): p is ObjectProperty =>
          p.type === 'ObjectProperty'
          && p.key.type === 'Identifier'
          && p.key.name === 'name',
      )?.value
      if (nameNode?.type !== 'StringLiteral') continue
      const name = nameNode.value
      if (!name || name.startsWith('_')) continue
      results.push({
        rawName: name,
        replace: (resolved) => {
          s.prependLeft(renderStart!, `\nthis.$options.directives["${name}"] = ${resolved};`)
        },
      })
    }
  }

  return results
}

const resolveVue3 = (code: string, s: MagicString): ResolveResult[] => {
  const results: ResolveResult[] = []

  for (const match of code.matchAll(/_resolveDirective\("(.+?)"\)/g)) {
    const matchedName = match[1]
    if (match.index != null && matchedName && !matchedName.startsWith('_')) {
      const start = match.index
      const end = start + match[0].length
      results.push({
        rawName: matchedName,
        replace: resolved => s.overwrite(start, end, resolved),
      })
    }
  }

  return results
}

export default async(code: string, version: VueVersion, s: MagicString, ctx: Context, sfcPath: string) => {
  let no = 0

  const results = version === 'vue2' ? resolveVue2(code, s) : resolveVue3(code, s)
  for (const { rawName, replace } of results) {
    debug(`| ${rawName}`)
    const name = pascalCase(rawName)
    ctx.updateUsageMap(sfcPath, [name])

    const directive = await ctx.findComponent(name, 'directive', [sfcPath])
    if (!directive) continue

    const varName = `__unplugin_directives_${no}`
    s.prepend(`${stringifyComponentImport({ ...directive, name: varName }, ctx)};\n`)
    no += 1
    replace(varName)
  }

  debug(`^ (${no})`)
}
