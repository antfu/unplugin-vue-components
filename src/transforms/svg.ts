import { readFileSync } from 'fs'
import Debug from 'debug'
import type { Transform } from 'vite'
import { compileTemplate } from '@vue/compiler-sfc'
import { Context } from '../context'

const debug = Debug('vite-plugin-components:transform:svg')

/**
 * @param ctx
 */
export function SVGTransformer(ctx: Context): Transform {
  return {
    test({ path, query }) {
      return path.endsWith('.svg') && query.import != null
    },
    transform({ path }) {
      debug(path)
      const svg = readFileSync(path, 'utf8')
      let { code } = compileTemplate({
        source: svg,
        transformAssetUrls: false,
      })
      code = code.replace('export function render', 'export default function render')
      return `${code}`
    },
  }
}
