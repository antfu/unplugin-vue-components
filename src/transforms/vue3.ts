import Debug from 'debug'
import { Transformer } from '../types'
import { Context } from '../context'
import { pascalCase, stringifyComponentImport } from '../utils'

const debug = Debug('vite-plugin-components:transform:vue3')

export function Vue3Transformer(ctx: Context): Transformer {
  return (code, id, path, query) => {
    if (!(path.endsWith('.vue') || ctx.options.customLoaderMatcher(id)))
      return code

    ctx.searchGlob()

    const sfcPath = ctx.normalizePath(path)
    debug(sfcPath)

    const head: string[] = []
    let no = 0
    const componentPaths: string[] = []

    let transformed = code.replace(/_resolveComponent\("(.+?)"\)/g, (str, match) => {
      if (match && !match.startsWith('_')) {
        debug(`| ${match}`)
        const name = pascalCase(match)
        componentPaths.push(name)
        const component = ctx.findComponent(name, [sfcPath])
        if (component) {
          const var_name = `__vite_components_${no}`
          if (ctx.options.importPathTransform) {
            const result = ctx.options.importPathTransform(component.path)
            if (result != null)
              component.path = result
          }

          if (component.importName)
            head.push(`import { ${component.importName} as ${var_name} } from '${component.path}'`)
          else
            head.push(`import ${var_name} from '${component.path}'`)

          if (component.stylePath)
            head.push(`import '${component.stylePath}'`)

          no += 1
          return var_name
        }
      }
      return str
    })

    debug(`^ (${no})`)

    ctx.updateUsageMap(sfcPath, componentPaths)

    transformed = `${head.join('\n')}\n${transformed}`

    return transformed
  }
}
