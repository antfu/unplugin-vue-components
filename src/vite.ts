import type { Plugin } from 'vite'
import type { ComponentsApi, Options } from './types'
import unplugin from '.'

export default unplugin.vite as (options?: Options | undefined) => Plugin & { api: ComponentsApi }
