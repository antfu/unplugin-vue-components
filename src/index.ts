import type { Plugin } from 'vite'
import { createRollupPlugin } from './plugins/build'
import { createServerPlugin } from './plugins/server'
import { Options } from './types'
import { VueScriptTransformer } from './transforms/vueScript'
import { VueTemplateTransformer } from './transforms/vueTemplate'
import { SVGTransformer } from './transforms/svg'
import { Context } from './context'

const defaultOptions: Options = {
  dirs: 'src/components',
  extensions: 'vue',
  deep: true,

  directoryAsNamespace: false,
  globalNamespaces: [],

  alias: {},
  root: process.cwd(),
}

function VitePluginComponents(options: Partial<Options> = {}): Plugin {
  const resolvedOptions: Options = Object.assign({}, defaultOptions, options)
  const ctx: Context = new Context(resolvedOptions)

  return {
    configureServer: createServerPlugin(ctx),
    rollupInputOptions: {
      plugins: [
        createRollupPlugin(ctx),
      ],
    },
    transforms: [
      VueScriptTransformer(ctx),
      VueTemplateTransformer(ctx),
      SVGTransformer(ctx),
    ],
  }
}

export type { Options }
export default VitePluginComponents
