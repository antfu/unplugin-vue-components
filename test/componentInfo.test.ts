import { resolve } from 'path'
import { Context } from '../src/context'
import { addImports } from '../src/transforms/vue3'

describe('component info', () => {
  it('stylePath', async() => {
    const ctx = new Context({}, { root: resolve(__dirname, '../examples/vue3') } as any)
    const result: Array<string> = []
    addImports(ctx, {
      name: 'Test',
      path: 'test',
      stylePath: 'test.css',
    }, result, 'var1')
    expect(result).toMatchSnapshot()
  })
  it('stylePath as array', async() => {
    const ctx = new Context({}, { root: resolve(__dirname, '../examples/vue3') } as any)
    const result: Array<string> = []
    addImports(ctx, {
      name: 'Test',
      path: 'test',
      stylePath: ['test1.css', 'test2.css'],
    }, result, 'var1')
    expect(result).toMatchSnapshot()
  })
  it('importName', async() => {
    const ctx = new Context({}, { root: resolve(__dirname, '../examples/vue3') } as any)
    const result: Array<string> = []
    addImports(ctx, {
      name: 'Test',
      path: 'test',
      importName: 'customImportName',
    }, result, 'var1')
    expect(result).toMatchSnapshot()
  })
})
