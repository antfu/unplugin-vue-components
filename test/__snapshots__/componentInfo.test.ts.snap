// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component info importName 1`] = `
Array [
  "import { customImportName as var1 } from 'test'",
]
`;

exports[`component info stylePath 1`] = `
Array [
  "import var1 from 'test'",
  "import 'test.css'",
]
`;

exports[`component info stylePath as array 1`] = `
Array [
  "import var1 from 'test'",
  "import 'test1.css'",
  "import 'test2.css'",
]
`;
